csrc/mmdeploy/net/ipu/ipu_net.cpp

```diff
@@ -0,0 +1,193 @@
+    *(to_ptr + i) = pfrom[i];
+  }
+}
+
+// IPUNet::IPUNet() {}
+IPUNet::~IPUNet() {}
+
+Result<void> IPUNet::Init(const Value& args) {
+  
+  std::string popef_path = args["popef_path"].get<std::string>();
+
+  batch_per_step = args["batches_per_step"].get<int>();
+
+  mconfig.device_wait_config =
+      model_runtime::DeviceWaitConfig(std::chrono::seconds{600}, std::chrono::seconds{1});
+  model_runner = std::make_unique<model_runtime::ModelRunner>(popef_path, mconfig);
+
+  input_desc = model_runner->getExecuteInputs();
+  output_desc = model_runner->getExecuteOutputs();
+
+  for (int i=0; i<input_desc.size(); i++){
+    input_desc[i].shape[0] *= batch_per_step;
+    input_desc[i].size_in_bytes *= batch_per_step;
+    MMDEPLOY_INFO("input desc shape {} ", input_desc[i].shape);
+  } 
+
+  for (int i=0; i<output_desc.size(); i++){
+    output_desc[i].shape[0] *= batch_per_step;
+    output_desc[i].size_in_bytes *= batch_per_step;
+  } 
+
+  input_memory = examples::allocateHostInputData(input_desc);
+  output_memory = examples::allocateHostInputData(output_desc);
+
+  for (int i = 0; i < input_desc.size(); i++) {
+    auto desc = input_desc[i];
+    MMDEPLOY_INFO("input desc dtype {} ", desc.data_type);
+
+    mmdeploy::DataType dtype = ipu_type_convert(desc.data_type);
+
+    input_tensors_.emplace_back(TensorDesc{
+        Device("cpu"),
+        dtype,
+        desc.shape,
+        desc.name,
+    });
+  }
+
+  for (int i = 0; i < output_desc.size(); i++) {
+    auto desc = output_desc[i];
+    MMDEPLOY_INFO("output desc dtype {} ", desc.data_type);
+    mmdeploy::DataType dtype = ipu_type_convert(desc.data_type);
+
+    output_tensors_.emplace_back(TensorDesc{
+        Device("cpu"),
+        dtype,
+        desc.shape,
+        desc.name,
+    });
+  }
+  return success();
+}
+
+Result<void> IPUNet::Deinit() { return success(); }
+
+Result<void> IPUNet::Reshape(Span<TensorShape> input_shapes) {
+  for (size_t i = 0; i < input_shapes.size(); ++i) {
+    input_tensors_[i].Reshape(input_shapes[i]);
+  }
+  return success();
+}
+
+Result<Span<Tensor>> IPUNet::GetInputTensors() { return input_tensors_; }
+
+Result<Span<Tensor>> IPUNet::GetOutputTensors() { return output_tensors_; }
+
+Result<void> IPUNet::Forward() {
+  MMDEPLOY_INFO("ipu device running forward ");
+  {
+    // copy input to itensor buffer
+
+    for (auto& tensor : input_tensors_) {
+      const auto& name = tensor.desc().name;
+      copy_input(tensor, input_memory[name]);
+    }
+  }
+
+  {
+    struct timeval start, end;
+    gettimeofday(&start, 0);
+    model_runner->execute(examples::toInputMemoryView(input_memory),
+                          examples::toOutputMemoryView(output_memory));
+    gettimeofday(&end, 0);
+    long sec = end.tv_sec - start.tv_sec;
+    long micro = end.tv_usec - start.tv_usec;
+    double elapsed = sec + micro * 1e-6;
+    
+    MMDEPLOY_INFO("ipu inference done, time elapsed {} sec", elapsed);
```

Comment:
no need to count time here. We got a feature named `profile` in SDK. Once enabled, SDK will evaluate time automatically.